File: backend-java/src/main/java/com/smartsupply/service/AiService.java
package com.smartsupply.service;

import com.smartsupply.entity.*;
import com.smartsupply.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.springframework.transaction.annotation.Transactional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class AiService {

    private final ProductRepository productRepository;
    private final InventoryItemRepository inventoryRepository;
    private final SupplierRepository supplierRepository;
    private final WarehouseRepository warehouseRepository;
    private final PurchaseOrderRepository purchaseOrderRepository;
    private final InventoryMovementRepository inventoryMovementRepository;

    @Value("${gemini.api-key}") // Fix: Ensure this matches application.yml key, often camelCase or kebab-case
    private String geminiApiKey;

    private final RestTemplate restTemplate = new RestTemplate();
    private static final String GEMINI_API_URL = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-lite:generateContent?key=";

    @Transactional(readOnly = true)
    public String generateResponse(String userMessage) {
        try {
            // 1. Aggregate Context (Wrap in try-catch to handle DB errors)
            String context = buildContext();

            // 2. Build Prompt
            String systemPrompt = "You are SmartSupply Assistant, an AI expert in supply chain management. " +
                    "Use the provided database context to answer the user's question. " +
                    "You SHOULD aggregate, summarize, and count data when asked (e.g., 'total inventory', 'how many products'). " +
                    "If the answer is not in the data, say you don't know. " +
                    "Be concise but informative. Format money as EUR (€). " +
                    "\n\nDATA CONTEXT:\n" + context;

            // 3. Call Gemini
            return callGeminiApi(systemPrompt, userMessage);
            
        } catch (Exception e) {
            log.error("Critical Error in AiService", e);
            // Return error as a chat message so frontend doesn't break
            return "I am experiencing technical difficulties. Detailed error: " + e.getMessage();
        }
    }

    private String buildContext() {
        StringBuilder sb = new StringBuilder();

        // Products
        List<Product> products = productRepository.findAll();
        sb.append("PRODUCTS (Total: ").append(products.size()).append("):\n");
        for (Product p : products) {
            sb.append(String.format("- %s (SKU: %s, Price: %.2f, Category: %s, Safety Stock: %d)\n",
                    p.getName(), p.getSku(), p.getPrice(), p.getCategory(), (p.getSafetyStock() != null ? p.getSafetyStock() : 0)));
        }
        sb.append("\n");

        // Warehouses
        List<Warehouse> warehouseList = warehouseRepository.findAll();
        sb.append("WAREHOUSES:\n");
        for (Warehouse w : warehouseList) {
            sb.append(String.format("- %s (%s)\n", w.getName(), w.getLocation()));
        }
        sb.append("\n");

        // Inventory
        List<InventoryItem> inventory = inventoryRepository.findAll();
        int totalQuantity = inventory.stream().mapToInt(InventoryItem::getQuantity).sum();
        double totalValue = inventory.stream()
            .mapToDouble(i -> i.getQuantity() * i.getProduct().getPrice().doubleValue())
            .sum();
            
        sb.append(String.format("INVENTORY SUMMARY: Total Items: %d, Total Value: %.2f EUR\n", totalQuantity, totalValue));
        sb.append("INVENTORY DETAILS:\n");
        for (InventoryItem i : inventory) {
             int safetyStock = i.getProduct().getSafetyStock() != null ? i.getProduct().getSafetyStock() : 0;
             boolean isLowStock = i.getQuantity() <= safetyStock;
             
             sb.append(String.format("- Product: %s, Warehouse: %s, Quantity: %d%s\n",
                     i.getProduct().getName(), 
                     i.getWarehouse().getName(), 
                     i.getQuantity(),
                     (isLowStock ? " [LOW STOCK WARNING]" : "")));
        }
        sb.append("\n");

        // Suppliers
        List<Supplier> suppliers = supplierRepository.findAll();
        sb.append("SUPPLIERS:\n");
        for (Supplier s : suppliers) {
            sb.append(String.format("- %s (Contact: %s, Email: %s)\n",
                    s.getName(), s.getContactPerson(), s.getEmail()));
        }
        sb.append("\n");

        // Recent POs (Limit to last 10 for context size)
        List<PurchaseOrder> orders = purchaseOrderRepository.findAll(); // In real app, use Pageable
        sb.append("RECENT PURCHASE ORDERS:\n");
        for (PurchaseOrder po : orders) {
             String itemsSummary = po.getItems().stream()
                 .map(item -> String.format("%dx %s", item.getQuantityOrdered(), item.getProduct().getName()))
                 .collect(Collectors.joining(", "));
             
             sb.append(String.format("- Order #%s: Supplier: %s, Status: %s, Total: %.2f, Items: [%s]\n",
                     po.getId(), po.getSupplier().getName(), po.getStatus(), po.getTotalAmount(), itemsSummary));
        }
        sb.append("\n");

        // Recent Inventory Movements (History)
        List<InventoryMovement> movements = inventoryMovementRepository.findAll(); // Limit in real app
        sb.append("RECENT INVENTORY MOVEMENTS (History):\n");
        for (InventoryMovement m : movements) {
            sb.append(String.format("- %s: %s %d units of %s at %s (Reason: %s, User: %s)\n",
                    m.getCreatedAt(), m.getMovementType(), m.getQuantity(), 
                    m.getInventoryItem().getProduct().getName(),
                    m.getInventoryItem().getWarehouse().getName(),
                    m.getReason(),
                    (m.getPerformedBy() != null ? m.getPerformedBy().getEmail() : "System")));
        }

        return sb.toString();
    }

    private String callGeminiApi(String systemPrompt, String userMessage) {
        try {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);

            // Gemini Request Body Structure
            Map<String, Object> requestBody = new HashMap<>();
            
            // Gemini 1.5 format: contents: [{ role: "user", parts: [{ text: "..." }] }]
            // We prepend system prompt to the user message
            
            String combinedPrompt = systemPrompt + "\n\nUSER QUESTION: " + userMessage;
            
            Map<String, Object> part = new HashMap<>();
            part.put("text", combinedPrompt);

            Map<String, Object> content = new HashMap<>();
            content.put("parts", List.of(part));

            requestBody.put("contents", List.of(content));

            HttpEntity<Map<String, Object>> entity = new HttpEntity<>(requestBody, headers);

            // Use exchange with ParameterizedTypeReference to avoid raw type warnings
            ResponseEntity<Map<String, Object>> response = restTemplate.exchange(
                    GEMINI_API_URL + geminiApiKey,
                    org.springframework.http.HttpMethod.POST,
                    entity,
                    new ParameterizedTypeReference<Map<String, Object>>() {}
            );

            // Parse Response
            // Response structure: candidates[0].content.parts[0].text
            Map<String, Object> responseBody = response.getBody();
            if (responseBody != null && responseBody.containsKey("candidates")) {
                @SuppressWarnings("unchecked")
                List<Map<String, Object>> candidates = (List<Map<String, Object>>) responseBody.get("candidates");
                if (!candidates.isEmpty()) {
                    @SuppressWarnings("unchecked")
                    Map<String, Object> contentMap = (Map<String, Object>) candidates.get(0).get("content");
                    @SuppressWarnings("unchecked")
                    List<Map<String, Object>> parts = (List<Map<String, Object>>) contentMap.get("parts");
                    if (!parts.isEmpty()) {
                        return (String) parts.get(0).get("text");
                    }
                }
            }

            return "I'm sorry, I couldn't understand that.";

        } catch (Exception e) {
            log.error("Error calling Gemini API", e);
            throw new RuntimeException("Gemini API Error: " + e.getMessage());
        }
    }
}

File: backend-fastapi/ai_service.py
"""
AI Service — изпраща данни от базата + въпроса на потребителя към Google Gemini.

Как работи (RAG pattern):

    Потребител пита: "Кои продукти са с ниско количество?"
         ↓
    1. _build_context() → чете ВСИЧКИ данни от базата
       и ги форматира като текст (продукти, складове, количества...)
         ↓
    2. Комбинира текста с въпроса в един голям prompt
         ↓
    3. _call_gemini_api() → изпраща prompt-а към Google Gemini
         ↓
    4. Gemini чете данните и отговаря на въпроса
         ↓
    5. Връщаме отговора на потребителя

    Защо работи: Gemini не знае нищо за нашата база данни.
    Но когато му дадем данните като текст, той може да ги анализира
    и да отговори на въпроси за тях.
"""

import os
import logging
import requests
from collections import defaultdict
from sqlmodel import Session, select

from models import Product, Warehouse, Supplier, InventoryItem, InventoryMovement

logger = logging.getLogger(__name__)

GEMINI_API_URL = (
    "https://generativelanguage.googleapis.com/v1beta/"
    "models/gemini-2.5-flash-lite:generateContent"
)


def generate_response(user_message: str, session: Session) -> str:
    """
    Главна функция. Извиква се от main.py когато потребител изпрати чат съобщение.

    Параметри:
        user_message: текстът, който потребителят е написал
        session: връзка към базата данни (подаден от FastAPI)
    """
    try:
        # Стъпка 1: Прочети всички данни от базата и ги форматирай като текст
        context = _build_context(session)

        # Стъпка 2: Създай инструкции за AI-я (system prompt)
        # Казваме му: "Ти си асистент, ето ти данните, отговори на въпроса"
        system_prompt = (
            "You are SmartSupply Assistant, an AI expert in supply chain management. "
            "Use the provided database context to answer the user's question. "
            "You SHOULD aggregate, summarize, and count data when asked "
            "(e.g., 'total inventory', 'how many products'). "
            "If the answer is not in the data, say you don't know. "
            "Be concise but informative. Format money as EUR (€). "
            "\n\nDATA CONTEXT:\n" + context
        )

        # Стъпка 3: Изпрати към Gemini API и върни отговора
        return _call_gemini_api(system_prompt, user_message)

    except Exception as e:
        logger.error("AI Service error: %s", e, exc_info=True)
        return f"Technical error: {e}"


def _build_context(session: Session) -> str:
    """
    Чете ВСИЧКИ данни от базата и ги форматира като текст.

    Това е "Retrieval" частта от RAG:
    R = Retrieval (извличане на данни)
    A = Augmented (добавяне на данните към prompt-а)
    G = Generation (Gemini генерира отговор)

    Връща string от ~30,000 символа с всички продукти, складове,
    количества, доставчици и последни движения.
    """
    lines = []

    # ── PRODUCTS ──
    products = session.exec(select(Product)).all()
    products_by_id = {p.id: p for p in products}
    # ^ Създаваме dict за бърз достъп по id: {1: Product(...), 2: Product(...), ...}

    lines.append(f"PRODUCTS (Total: {len(products)}):")
    for p in products:
        lines.append(
            f"- {p.name} (SKU: {p.sku}, Price: {p.price:.2f}, "
            f"Category: {p.category}, Safety Stock: {p.safety_stock})"
        )
    lines.append("")

    # ── WAREHOUSES ──
    warehouses = session.exec(select(Warehouse)).all()
    warehouses_by_id = {w.id: w for w in warehouses}

    lines.append(f"WAREHOUSES (Total: {len(warehouses)}):")
    for w in warehouses:
        lines.append(f"- {w.name} ({w.location})")
    lines.append("")

    # ── INVENTORY SUMMARY ──
    items = session.exec(select(InventoryItem)).all()

    # sum() с generator expression:
    # sum(item.quantity for item in items)
    #   = вземи item.quantity от всеки item, и ги сумирай
    #   Като SUM(quantity) в SQL
    total_quantity = sum(item.quantity for item in items)
    total_value = sum(
        item.quantity * products_by_id[item.product_id].price
        for item in items
        if item.product_id in products_by_id
    )

    lines.append(
        f"INVENTORY SUMMARY: Total Items: {total_quantity}, "
        f"Total Value: {total_value:.2f} EUR"
    )

    # ── MOST STOCKED PRODUCT ──
    # defaultdict(int) = dict, който автоматично добавя 0 за нов ключ
    # Нормален dict: ако product_stock["Bolt"] не съществува → KeyError
    # defaultdict(int): ако product_stock["Bolt"] не съществува → създава го с 0
    product_stock = defaultdict(int)
    for item in items:
        product = products_by_id.get(item.product_id)
        if product:
            product_stock[product.name] += item.quantity

    if product_stock:
        # max() с key= → намери ключа с най-голяма стойност
        most_stocked = max(product_stock, key=product_stock.get)
        lines.append(
            f"MOST STOCKED PRODUCT: {most_stocked} "
            f"(Total: {product_stock[most_stocked]})"
        )
    lines.append("")

    # ── INVENTORY DETAILS (с предупреждения за ниско количество) ──
    lines.append("INVENTORY DETAILS:")
    sorted_items = sorted(items, key=lambda x: x.quantity, reverse=True)
    # sorted() с key=lambda → сортирай по quantity, от най-голямото
    # lambda x: x.quantity е кратък запис за "вземи quantity от x"
    # reverse=True → низходящо (500, 200, 50, 10...)

    for item in sorted_items:
        product = products_by_id.get(item.product_id)
        warehouse = warehouses_by_id.get(item.warehouse_id)
        if product and warehouse:
            low = " [LOW STOCK]" if item.quantity <= product.safety_stock else ""
            lines.append(f"- {product.name} at {warehouse.name}: {item.quantity}{low}")
    lines.append("")

    # ── SUPPLIERS ──
    suppliers = session.exec(select(Supplier)).all()
    lines.append(f"SUPPLIERS (Total: {len(suppliers)}):")
    for s in suppliers:
        lines.append(f"- {s.name} (Contact: {s.contact_person}, Email: {s.email})")
    lines.append("")

    # ── RECENT MOVEMENTS ──
    movements = session.exec(
        select(InventoryMovement).order_by(InventoryMovement.created_at.desc())
    ).all()[:100]
    # [:100] = вземи само първите 100 (не искаме да пратим 756 реда на AI)

    lines.append("RECENT INVENTORY MOVEMENTS (Last 100):")
    for m in movements:
        item_for_m = next((i for i in items if i.id == m.inventory_item_id), None)
        product_name = "Unknown"
        warehouse_name = "Unknown"
        if item_for_m:
            p = products_by_id.get(item_for_m.product_id)
            w = warehouses_by_id.get(item_for_m.warehouse_id)
            product_name = p.name if p else "Unknown"
            warehouse_name = w.name if w else "Unknown"
        lines.append(
            f"- {m.created_at}: {m.movement_type} {m.quantity} units of "
            f"{product_name} at {warehouse_name} (Reason: {m.reason})"
        )

    # "\n".join(lines) = свързва всички редове с нов ред между тях
    # ["А", "Б", "В"] → "А\nБ\nВ"
    return "\n".join(lines)


def _call_gemini_api(system_prompt: str, user_message: str) -> str:
    """
    Изпраща prompt към Google Gemini API и връща отговора.

    Gemini API-то приема JSON в определен формат и връща JSON.
    Ние ползваме requests библиотеката за HTTP заявки.
    """

    # Четем API ключа от .env файла
    api_key = os.getenv("GEMINI_API_KEY", "")
    if not api_key:
        return "Error: GEMINI_API_KEY is not set in .env file."

    # Комбинираме system prompt + въпроса на потребителя
    combined_prompt = f"{system_prompt}\n\nUSER QUESTION: {user_message}"

    # Тялото на заявката — в точния формат, който Gemini API очаква
    request_body = {"contents": [{"parts": [{"text": combined_prompt}]}]}

    # Изпращаме POST заявка към Gemini
    # json=request_body → requests автоматично:
    #   1. Конвертира dict-а в JSON string
    #   2. Добавя header Content-Type: application/json
    response = requests.post(
        f"{GEMINI_API_URL}?key={api_key}",
        json=request_body,
        timeout=30,
    )

    # Ако сървърът върне грешка (4xx, 5xx) → хвърли exception
    response.raise_for_status()

    # Парсваме JSON отговора
    # Gemini връща:
    # {
    #   "candidates": [
    #     {
    #       "content": {
    #         "parts": [
    #           {"text": "Отговорът на AI тук..."}
    #         ]
    #       }
    #     }
    #   ]
    # }
    data = response.json()

    # Навигираме до текста на отговора
    # .get() е безопасен достъп — ако ключът не съществува, връща default
    candidates = data.get("candidates", [])
    if candidates:
        parts = candidates[0].get("content", {}).get("parts", [])
        if parts:
            return parts[0].get("text", "Couldn't generate a response.")

    return "Couldn't understand the response from Gemini."
